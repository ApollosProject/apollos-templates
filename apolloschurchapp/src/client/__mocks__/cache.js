import {
  InMemoryCache,
  IntrospectionFragmentMatcher,
} from 'apollo-cache-inmemory';
import introspectionQueryResultData from '../fragmentTypes.json';

const nodeCacheRedirect = (_, { id }, { getCacheKey }) =>
  id ? getCacheKey({ __typename: id.split(':')[0], id }) : null;

const cache = new InMemoryCache({
  fragmentMatcher: new IntrospectionFragmentMatcher({
    introspectionQueryResultData,
  }),
  cacheRedirects: {
    Query: {
      node: nodeCacheRedirect,
    },
  },
});

export default cache;
