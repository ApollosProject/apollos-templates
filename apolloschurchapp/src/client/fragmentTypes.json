{"__schema":{"types":[{"kind":"INTERFACE","name":"Node","possibleTypes":[{"name":"ContentChannel"},{"name":"Person"},{"name":"Campus"},{"name":"Event"},{"name":"Group"},{"name":"Scripture"},{"name":"Pass"},{"name":"UniversalContentItem"},{"name":"DevotionalContentItem"},{"name":"MediaContentItem"},{"name":"ContentSeriesContentItem"},{"name":"WeekendContentItem"},{"name":"ActionListFeature"},{"name":"TextFeature"},{"name":"ScriptureFeature"}]},{"kind":"INTERFACE","name":"ContentItem","possibleTypes":[{"name":"UniversalContentItem"},{"name":"DevotionalContentItem"},{"name":"MediaContentItem"},{"name":"ContentSeriesContentItem"},{"name":"WeekendContentItem"}]},{"kind":"INTERFACE","name":"Media","possibleTypes":[{"name":"ImageMedia"},{"name":"VideoMedia"},{"name":"AudioMedia"}]},{"kind":"INTERFACE","name":"MediaSource","possibleTypes":[{"name":"ImageMediaSource"},{"name":"VideoMediaSource"},{"name":"AudioMediaSource"}]},{"kind":"INTERFACE","name":"Sharable","possibleTypes":[{"name":"SharableContentItem"},{"name":"SharableFeature"}]},{"kind":"INTERFACE","name":"Feature","possibleTypes":[{"name":"ActionListFeature"},{"name":"TextFeature"},{"name":"ScriptureFeature"}]}]}}