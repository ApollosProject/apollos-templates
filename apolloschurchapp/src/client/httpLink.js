import { Platform } from 'react-native';
import { createUploadLink } from 'apollo-upload-client';
import ApollosConfig from '@apollosproject/config';
import { split } from 'apollo-link';
import { createHttpLink } from 'apollo-link-http';
import { getMainDefinition } from 'apollo-utilities';

let uri = ApollosConfig.APP_DATA_URL;
const androidUri = ApollosConfig.ANDROID_URL || '10.0.2.2';

// Android's emulator requires localhost network traffic to go through 10.0.2.2
if (Platform.OS === 'android') uri = uri.replace('localhost', androidUri);

export default split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'mutation';
  },
  createUploadLink({ uri }),
  createHttpLink({
    uri,
    useGETForQueries: true,
  })
);
