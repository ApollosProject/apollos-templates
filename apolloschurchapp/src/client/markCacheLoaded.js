import gql from 'graphql-tag';

export default gql`
  mutation markCacheLoaded {
    cacheMarkLoaded @client
  }
`;
