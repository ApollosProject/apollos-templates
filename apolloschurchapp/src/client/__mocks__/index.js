import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import ApolloClient from 'apollo-client';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';
import { SchemaLink } from 'apollo-link-schema';
import { importSchema } from 'graphql-import';

import cache from '../cache';
import { resolvers, schema, defaults } from '../../store';

// eslint-disable-next-line
export default (props) => {
  let finalProps = props;
  // eslint-disable-next-line
  if (!props.mocks) {
    // eslint-disable-next-line
    finalProps = { ...props, resolvers };
  }
  return <MockedProvider cache={cache} {...finalProps} />;
};

const serverSchema = makeExecutableSchema({
  typeDefs: importSchema(
    `${process.cwd()}/../apollos-church-api/local.graphql`
  ),
  resolverValidationOptions: {
    requireResolversForResolveType: false,
  },
});
addMockFunctionsToSchema({ schema: serverSchema });

const link = new SchemaLink({ schema: serverSchema });
cache.writeData({ data: defaults });

export const client = new ApolloClient({
  link,
  cache,
  resolvers,
  typeDefs: schema,
});
