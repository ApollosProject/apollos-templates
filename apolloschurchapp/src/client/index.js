import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { getVersion, getApplicationName } from 'react-native-device-info';

import { authLink, buildErrorLink } from '@apollosproject/ui-auth';
import { resolvers, schema, defaults } from '../store';
import NavigationService from '../NavigationService';
import httpLink from './httpLink';
import cache, { ensureCacheHydration, MARK_CACHE_LOADED } from './cache';

const goToAuth = () => NavigationService.resetToAuth();
const wipeData = () => cache.writeData({ data: defaults });

let clearStore;
let storeIsResetting = false;
const onAuthError = async () => {
  if (!storeIsResetting) {
    storeIsResetting = true;
    await clearStore();
  }
  storeIsResetting = false;
  goToAuth();
};

const errorLink = buildErrorLink(onAuthError);

const link = ApolloLink.from([authLink, errorLink, httpLink]);

export const client = new ApolloClient({
  link,
  cache,
  queryDeduplication: false,
  shouldBatch: true,
  resolvers,
  typeDefs: schema,
  name: getApplicationName(),
  version: getVersion(),
});

// Hack to give auth link access to method on client;
// eslint-disable-next-line prefer-destructuring
clearStore = client.clearStore;

wipeData();
// Ensure that media player still works after logout.
client.onClearStore(() => wipeData());

class ClientProvider extends PureComponent {
  static propTypes = {
    client: PropTypes.shape({
      cache: PropTypes.shape({}),
    }),
  };

  static defaultProps = {
    client,
  };

  async componentDidMount() {
    try {
      await ensureCacheHydration;
    } catch (e) {
      throw e;
    } finally {
      client.mutate({ mutation: MARK_CACHE_LOADED });
    }
  }

  render() {
    return <ApolloProvider {...this.props} client={client} />;
  }
}

export default ClientProvider;
