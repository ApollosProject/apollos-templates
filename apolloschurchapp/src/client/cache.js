import { InMemoryCache } from '@apollo/client/cache';
import AsyncStorage from '@react-native-community/async-storage';
import { CachePersistor } from 'apollo3-cache-persist';
import ApollosConfig from '@apollosproject/config';

import introspectionQueryResultData from './fragmentTypes.json';

const finalPossibleTypes = {};
introspectionQueryResultData.__schema.types.forEach((supertype) => {
  if (supertype.possibleTypes) {
    finalPossibleTypes[supertype.name] = [
      ...supertype.possibleTypes.map((subtype) => subtype.name),
    ];
  }
});
// We reset our apollo cache based an env value and static number.
// In the future, we should also look at resetting the app when an error occurs related to Apollo.
// You can also increment this number to force a manual reset of the cache.
const SCHEMA_VERSION = `${ApollosConfig.SCHEMA_VERSION}-1`; // Must be a string.
const SCHEMA_VERSION_KEY = 'apollo-schema-version';

const nodeCacheRedirect = (_, { id }, { getCacheKey }) =>
  id ? getCacheKey({ __typename: id.split(':')[0], id }) : null;

const cache = new InMemoryCache({
  possibleTypes: finalPossibleTypes,
  cacheRedirects: {
    Query: {
      node: nodeCacheRedirect,
    },
  },
});

const persistor = new CachePersistor({
  cache,
  storage: AsyncStorage,
});

// 1. If we are on the same schema version, restore the cache.
// 2. If that fails, purge the cache and update the stored version so we don't try and restore agian
// 3. If we are on a new schema version, purge the cache.
// 4. If purging or setting AsyncStorage keys fails (or anything else) we capture the error and log it.

export const ensureCacheHydration = (async () => {
  // We wrap everything in try/catch because crashing on a cache restore is bad
  // ðŸ˜¬
  try {
    const currentVersion = await AsyncStorage.getItem(SCHEMA_VERSION_KEY);
    if (currentVersion === SCHEMA_VERSION) {
      // If the current version matches the latest version,
      // we're good to go and can restore the cache.
      try {
        await persistor.restore();
      } catch (restoreError) {
        // If the restore fails, we want to do our best to purge the cache.
        await persistor.purge();
        console.error('Error restoring cache, purging the cache', restoreError);
      }
    } else {
      // Otherwise, we'll want to purge the outdated persisted cache
      // and mark ourselves as having updated to the latest version.
      await persistor.purge();
      await AsyncStorage.setItem(SCHEMA_VERSION_KEY, SCHEMA_VERSION);
    }
  } catch (error) {
    console.error('Error restoring or purging Apollo cache', error);
  }
})();

export default cache;
