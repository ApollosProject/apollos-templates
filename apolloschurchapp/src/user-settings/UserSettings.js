import React from 'react';
import { ScrollView } from 'react-native';
import { useQuery, useMutation, useApolloClient, gql } from '@apollo/client';
import { useNavigation } from '@react-navigation/native';
import { getVersion, getBuildNumber } from 'react-native-device-info';
import { get } from 'lodash';

import {
  BackgroundView,
  TableView,
  Cell,
  CellIcon,
  CellText,
  Divider,
  Touchable,
  ActivityIndicator,
  NavigationService,
  H3,
  H4,
  H6,
  PaddedView,
  withTheme,
  styled,
} from '@apollosproject/ui-kit';
import {
  checkOnboardingStatusAndNavigate,
  onboardingComplete,
} from '@apollosproject/ui-onboarding';
import { GET_LOGIN_STATE, LOGOUT } from '@apollosproject/ui-auth';
import {
  RockAuthedWebBrowser,
  UserAvatarUpdate,
} from '@apollosproject/ui-connected';

const StyledPaddedView = withTheme(({ theme }) => ({
  style: {
    paddingLeft: theme.sizing.baseUnit,
    paddingBottom: theme.sizing.baseUnit * 0.5,
    paddingTop: 0,
  },
}))(PaddedView);

const StyledCellIcon = withTheme(({ theme }) => ({
  fill: theme.colors.neutral.gray2,
}))(CellIcon);

const Container = styled({
  alignItems: 'center',
  justifyContent: 'center',
})(PaddedView);

const UserSettings = () => {
  const navigation = useNavigation();
  const {
    data: { isLoggedIn = false },
    loading,
  } = useQuery(GET_LOGIN_STATE, { fetchPolicy: 'cache-and-network' });
  const [logout] = useMutation(LOGOUT);
  const client = useApolloClient();
  const { data } = useQuery(gql`
    query currentUserId {
      currentUser {
        id
        profile {
          firstName
          lastName
        }
      }
    }
  `);

  const firstName = get(data, 'currentUser.profile.firstName');
  const lastName = get(data, 'currentUser.profile.lastName');

  if (loading) return <ActivityIndicator />;
  if (!isLoggedIn) return null;
  return (
    <BackgroundView>
      <ScrollView>
        <Container>
          <UserAvatarUpdate />
          <H3>{firstName && lastName ? `${firstName} ${lastName}` : ''}</H3>
        </Container>
        <RockAuthedWebBrowser>
          {(openUrl) => (
            <>
              <TableView>
                <Touchable
                  onPress={() => {
                    navigation.navigate('LikedContentFeedConnected');
                  }}
                >
                  <Cell>
                    <StyledCellIcon name="like" />
                    <CellText>Likes</CellText>
                  </Cell>
                </Touchable>
              </TableView>
              <StyledPaddedView>
                <H6>Your Profile</H6>
              </StyledPaddedView>
              <TableView>
                <Touchable
                  onPress={() => {
                    navigation.navigate('PersonalDetails');
                  }}
                >
                  <Cell>
                    <CellText>Personal Details</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Touchable
                  onPress={() => {
                    navigation.navigate('Location');
                  }}
                >
                  <Cell>
                    <CellText>Location</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Touchable
                  onPress={() => {
                    navigation.navigate('ChangePassword');
                  }}
                >
                  <Cell>
                    <CellText>Change Password</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Touchable
                  onPress={() => {
                    navigation.navigate('Notifications');
                  }}
                >
                  <Cell>
                    <CellText>Notification Settings</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
              </TableView>
              <StyledPaddedView>
                <H6>Help & Feedback</H6>
              </StyledPaddedView>
              <TableView>
                <Touchable
                  onPress={() => {
                    openUrl('mailto:support@apollos.app');
                  }}
                >
                  <Cell>
                    <CellText>Give Feedback</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Touchable
                  onPress={() => openUrl('https://apollosrock.newspring.cc/')}
                >
                  <Cell>
                    <CellText>Privacy Policy</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Touchable
                  onPress={() => openUrl('https://apollosrock.newspring.cc/')}
                >
                  <Cell>
                    <CellText>Terms of Use</CellText>
                    <StyledCellIcon name="arrow-next" />
                  </Cell>
                </Touchable>
                <Divider />
                <Cell>
                  <CellText>
                    {`App Version: ${getVersion()}.${getBuildNumber()}`}
                  </CellText>
                </Cell>
              </TableView>
              <TableView>
                <Touchable
                  onPress={async () => {
                    await logout();
                    // This resets the navigation stack, and the navigates to the first auth screen.
                    // This ensures that user isn't navigated to a subscreen of Auth, like the pin entry screen.
                    NavigationService.resetToAuth();
                  }}
                >
                  <Cell>
                    {
                      // Uncomment this once this icon is merged to master in core
                      // <StyledCellIcon name="arrow-down-right" />
                    }
                    <CellText>Logout</CellText>
                  </Cell>
                </Touchable>
              </TableView>

              {/* testing panel */}
              {process.env.NODE_ENV !== 'production' ? (
                <>
                  <PaddedView>
                    <H4>For development only</H4>
                  </PaddedView>
                  <TableView>
                    <Touchable
                      onPress={() =>
                        checkOnboardingStatusAndNavigate({
                          latestOnboardingVersion: 2,
                          navigation: NavigationService,
                          client,
                        })
                      }
                    >
                      <Cell>
                        <CellText>Launch Onboarding</CellText>
                      </Cell>
                    </Touchable>
                    <Divider />
                    <Touchable
                      onPress={() =>
                        onboardingComplete({
                          version: 0,
                          userId: data?.currentUser?.id,
                        })
                      }
                    >
                      <Cell>
                        <CellText>Reset Onboarding to Unseen</CellText>
                      </Cell>
                    </Touchable>
                    <Divider />
                    <Touchable
                      onPress={() =>
                        onboardingComplete({
                          version: 1,
                          userId: data?.currentUser?.id,
                        })
                      }
                    >
                      <Cell>
                        <CellText>Reset Onboarding to Seen v1</CellText>
                      </Cell>
                    </Touchable>
                  </TableView>
                </>
              ) : null}
            </>
          )}
        </RockAuthedWebBrowser>
      </ScrollView>
    </BackgroundView>
  );
};

export default UserSettings;
