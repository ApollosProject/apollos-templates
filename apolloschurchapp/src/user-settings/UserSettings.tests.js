import React from 'react';

import { GET_LOGIN_STATE } from '@apollosproject/ui-auth';
import { renderWithApolloData, Providers } from '@apollosproject/ui-test-utils';
import { MockedProvider } from '@apollo/client/testing';
import { InMemoryCache } from '@apollo/client/cache';

import UserSettings from '.';

describe('UserSettings component', () => {
  it('renders UserSettings when logged in', async () => {
    const cache = new InMemoryCache();
    cache.writeQuery({
      query: GET_LOGIN_STATE,
      data: { isLoggedIn: true },
    });
    const navigation = { navigate: jest.fn() };
    const tree = await renderWithApolloData(
      <Providers cache={cache} MockedProvider={MockedProvider}>
        <UserSettings navigation={navigation} />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
});
