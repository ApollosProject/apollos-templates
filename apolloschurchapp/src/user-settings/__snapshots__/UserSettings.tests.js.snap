// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings component renders UserSettings when logged in 1`] = `null`;
